### ISSUE TITLE
Equivalent of caffe iter_size in Keras

### ISSUE THREAD

**Issue by paragon00** [keras-team/keras#8334]
Apologies if this is not the best place to ask this, and thanks for the great work thats been done in Keras!

Can we get the equivalent of caffe's iter_size parameter in Keras? This accumulates gradient calcs over several GPU cycles before doing the weight update. It effectively allows a larger batch size. TensorFlow doesn't natively have this but some ppl seem to have implemented it themselves, e.g 

https://stackoverflow.com/questions/42156957/how-to-update-model-parameters-with-accumulated-gradients

I think it'd be a useful parameter to have for model training in Keras ... or is there some easy way to implement this functionality already?

Shaun

**Comment by ozabluda** [comment]
I agree that this is a needed feature, not sure if it makes sense to introduce it into Keras before TF (maybe depends on TF plans for it). But Caffe `iter_size` was an API hack. It would be much better for existing `batch size` to remain a clear (logical) `batch size`, with an optimization hint for the physical `batch size`. Which, hopefully, eventually be determined automatically, depending on CPU/GPU, train/test, multi-GPU, distributed/non-distributed, which cuda/eigen/tf kernels are faster (depends on available ram, etc) etc, etc.

**Comment by Sucran** [comment]
I think this can be solve by writing an optimizer by yourself, some useful code can be found in #3556 , @ozabluda, will this feature be add to Keras?  or it have already add into Keras?

**Comment by ozabluda** [comment]
I think an advice "write an optimizer by yourself" hardly counts as resolution of the issue.

**Comment by Sucran** [comment]
@ozabluda Thank you for the suggestion.
I think this issue essentially equal to #3556,  writing an optimizer to support accumulates gradient calcs over several GPU cycles, which means updating the parameters after several steps, right?
I agree this solution -- "write an optimizer by yourself" -- hardly counts as a good idea of this issue, but it provied a 'easy' way to implement this functionality.

**Comment by mvoelk** [comment]
#5244 same issue...

**Comment by ppwwyyxx** [comment]
I wrote this optimizer a year ago - it takes an arbitrary optimizer and make it accumulate gradients:

https://github.com/ppwwyyxx/tensorpack/blob/099975c56b7e63d41da1da57549fedf8bd68084f/tensorpack/tfutils/optimizer.py#L132-L201

Simply copying the relevant code and dependencies should be enough to make it work in keras.
